Following patches address issues when the command line exceed 256 characters.
Patches provided by upstream maitainer Michael Albinus.
They have been added to mainstream tramp gate and will
appear with Emacs 30.2 and Tramp 2.7.2.4 on GNU ELPA.

https://lists.gnu.org/archive/html/tramp-devel/2025-04/msg00001.html
https://lists.gnu.org/archive/html/tramp-devel/2025-05/msg00017.html

diff -r b1a28f8a774d lisp/net/tramp-sh.el
--- a/lisp/net/tramp-sh.el	Tue May 13 05:05:01 2025 -0700
+++ b/lisp/net/tramp-sh.el	Tue May 13 05:19:00 2025 -0700
@@ -643,6 +643,14 @@
 Format specifiers are replaced by `tramp-expand-script', percent
 characters need to be doubled.")
 
+(defconst tramp-readlink-file-truename
+  "if %s -h \"$1\"; then echo t; else echo nil; fi
+%s --canonicalize-missing \"$1\""
+  "Shell script to produce output suitable for use with `file-truename'
+on the remote file system.
+Format specifiers are replaced by `tramp-expand-script', percent
+characters need to be doubled.")
+
 (defconst tramp-perl-file-truename
   "%p -e '
 use File::Spec;
@@ -1286,15 +1294,13 @@
     (cond
      ;; Use GNU readlink --canonicalize-missing where available.
      ((tramp-get-remote-readlink v)
+      (tramp-maybe-send-script
+       v (format tramp-readlink-file-truename
+		 (tramp-get-test-command v) (tramp-get-remote-readlink v))
+       "tramp_readlink_file_truename")
       (tramp-send-command-and-check
-       v (format
-	  (concat
-	   "(if %s -h \"%s\"; then echo t; else echo nil; fi) && "
-	   "%s --canonicalize-missing %s")
-	  (tramp-get-test-command v)
-	  (tramp-shell-quote-argument localname)
-	  (tramp-get-remote-readlink v)
-	  (tramp-shell-quote-argument localname)))
+       v (format "tramp_readlink_file_truename %s"
+		 (tramp-shell-quote-argument localname)))
       (with-current-buffer (tramp-get-connection-buffer v)
 	(goto-char (point-min))
 	(tramp-set-file-property v localname "file-symlink-marker" (read (current-buffer)))
@@ -3096,7 +3102,9 @@
 			  (setq p (tramp-get-connection-process v))
 			  (process-put p 'remote-pid pid)
 			  (tramp-set-connection-property p "remote-pid" pid))
-			(when (memq connection-type '(nil pipe))
+			(when
+                            (or (memq connection-type '(nil pipe))
+                                (tramp-check-remote-uname v tramp-sunos-unames))
 			  ;; Disable carriage return to newline
 			  ;; translation.  This does not work on
 			  ;; macOS, see Bug#50748.
@@ -3110,8 +3118,11 @@
 			  ;; character to read.  When a process does
 			  ;; not read from stdin, like magit, it
 			  ;; should set a timeout
-			  ;; instead. See`tramp-pipe-stty-settings'.
+			  ;; instead.  See `tramp-pipe-stty-settings'.
 			  ;; (Bug#62093)
+			  ;; On Solaris, the maximum line length
+			  ;; depends also on MAX_CANON (256).  So we
+			  ;; disable buffering as well.
 			  ;; FIXME: Shall we rather use "stty raw"?
 			  (tramp-send-command
 			   v (format
